include/hc_printf.hpp

```diff
@@ -111,56 +117,91 @@ static inline void copy_n(char* dest, const char* src, unsigned int len) [[hc,cp
+  if (!string_buffer || soffset + str_len + 1 > queue[PRINTF_STRING_BUFFER_SIZE].data.ui){
     return PRINTF_STRING_BUFFER_OVERFLOW;
   }
   copy_n(&string_buffer[sb_offset], string, str_len + 1);
-  queue[offset].set(&string_buffer[sb_offset]);
+  queue[poffset].set(&string_buffer[sb_offset]);
+  soffset += str_len + 1;
   return PRINTF_SUCCESS;
 }
 
 template <typename T>
-static inline PrintfError set_batch(PrintfPacket* queue, int offset, const T t) [[hc,cpu]] {
+static inline PrintfError set_batch(PrintfPacket* queue, int poffset, unsigned int& soffset, const T t) [[hc,cpu]] {
   PrintfError err = PRINTF_SUCCESS;
-  queue[offset].set(t);
-  if (queue[offset].type == PRINTF_CHAR_PTR || queue[offset].type == PRINTF_CONST_CHAR_PTR){
-    err = process_str_batch(queue, offset, (char*)t);
+  queue[poffset].set(t);
+  if (queue[poffset].type == PRINTF_CHAR_PTR || queue[poffset].type == PRINTF_CONST_CHAR_PTR){
+    err = process_str_batch(queue, poffset, soffset, (char*)t);
   }
   return err;
 }
 template <typename T, typename... Rest>
-static inline PrintfError set_batch(PrintfPacket* queue, int offset, const T t, Rest... rest) [[hc,cpu]] {
+static inline PrintfError set_batch(PrintfPacket* queue, int poffset, unsigned int& soffset, const T t, Rest... rest) [[hc,cpu]] {
   PrintfError err = PRINTF_SUCCESS;
-  queue[offset].set(t);
-  if (queue[offset].type == PRINTF_CHAR_PTR || queue[offset].type == PRINTF_CONST_CHAR_PTR){
-    if ((err = process_str_batch(queue, offset, (char*)t)) != PRINTF_SUCCESS)
+  queue[poffset].set(t);
+  if (queue[poffset].type == PRINTF_CHAR_PTR || queue[poffset].type == PRINTF_CONST_CHAR_PTR){
+    if ((err = process_str_batch(queue, poffset, soffset, (char*)t)) != PRINTF_SUCCESS)
       return err;
   }
-  return set_batch(queue, offset + 1, rest...);
+  return set_batch(queue, poffset + 1, soffset, rest...);
 }
 
 template <typename... All>
 static inline PrintfError printf(PrintfPacket* queue, All... all) [[hc,cpu]] {
-  unsigned int count = 0;
-  countArg(count, all...);
+  unsigned int count_arg = 0;
+  unsigned int count_char = 0;
+  countArg(count_arg, count_char, all...);
 
   PrintfError error = PRINTF_SUCCESS;
+  PrintfPacketData old_off, try_off;
 
-  if (!queue || count + 1 + queue[1].data.ui > queue[0].data.ui) {
+  if (!queue || count_arg + 1 + queue[PRINTF_OFFSETS].data.uia[0] > queue[PRINTF_BUFFER_SIZE].data.ui) {
     error = PRINTF_BUFFER_OVERFLOW;
-  } else {
+  }
+  else if (!queue[PRINTF_STRING_BUFFER].data.ptr || count_char + queue[PRINTF_OFFSETS].data.uia[1] > queue[PRINTF_STRING_BUFFER_SIZE].data.ui){
+    error = PRINTF_STRING_BUFFER_OVERFLOW;
+  }
+  else {
+    do {
+      old_off.ull = queue[PRINTF_OFFSETS].data.al.load();
+      try_off.uia[0] = old_off.uia[0] + count_arg + 1;
+      try_off.uia[1] = old_off.uia[1] + count_char;
+    } while(!(queue[PRINTF_OFFSETS].data.al.compare_exchange_weak(old_off.ull, try_off.ull)));
```

Comment:
@aaronenyeshi this relates to my comments earlier in the file. I think the relationship between `ull`, `uia[]`, `al` deserved to be explained a bit better.